rustworkx-core/src/token_swapper.rs

```diff
@@ -0,0 +1,558 @@
+                            steps += 1;
+                            found = true;
+                            break;
+                        }
+                    }
+                    if !found {
+                        panic!("unexpected stop")
+                    }
+                }
+            }
+        }
+        if !todo_nodes.is_empty() {
+            panic!("got todo nodes");
+        }
+        swap_edges
+    }
+
+    fn swap(
+        &self,
+        node1: NodeIndex,
+        node2: NodeIndex,
+        digraph: &mut StableGraph<(), (), Directed>,
+        sub_digraph: &mut StableGraph<(), (), Directed>,
+        tokens: &mut HashMap<NodeIndex, NodeIndex>,
+        todo_nodes: &mut Vec<NodeIndex>,
+    ) {
+        // Get token values for the 2 nodes and remove them
+        let token1 = tokens.remove(&node1);
+        let token2 = tokens.remove(&node2);
+
+        // Swap the token edge values
+        if let Some(t2) = token2 {
+            tokens.insert(node1, t2);
+        }
+        if let Some(t1) = token1 {
+            tokens.insert(node2, t1);
+        }
+        // For each node, remove the (node, successor) from digraph and
+        // sub_digraph. Then add new token edges back in.
+        for node in [node1, node2] {
+            let edge_nodes: Vec<(NodeIndex, NodeIndex)> = digraph
+                .neighbors_directed(node, Outgoing)
+                .map(|successor| (node, successor))
+                .collect();
+            for (edge_node1, edge_node2) in edge_nodes {
+                let edge = digraph.find_edge(edge_node1, edge_node2).unwrap();
+                digraph.remove_edge(edge);
+            }
+            let edge_nodes: Vec<(NodeIndex, NodeIndex)> = sub_digraph
+                .neighbors_directed(node, Outgoing)
+                .map(|successor| (node, successor))
+                .collect();
+            for (edge_node1, edge_node2) in edge_nodes {
+                let edge = sub_digraph.find_edge(edge_node1, edge_node2).unwrap();
+                sub_digraph.remove_edge(edge);
+            }
+            self.add_token_edges(node, digraph, sub_digraph, tokens);
+
+            // If a node is a token key and not equal to the value, add it to todo_nodes
+            if tokens.contains_key(&node) && tokens[&node] != node {
+                if !todo_nodes.contains(&node) {
+                    todo_nodes.push(node);
+                }
+            // Otherwise if node is in todo_nodes, remove it
+            } else if todo_nodes.contains(&node) {
+                todo_nodes.swap_remove(todo_nodes.iter().position(|x| *x == node).unwrap());
+            }
+        }
+    }
+}
+
+/// Module to perform an approximately optimal Token Swapping algorithm. Supports partial
+/// mappings (i.e. not-permutations) for graphs with missing tokens.
+///
+/// Based on the paper: Approximation and Hardness for Token Swapping by Miltzow et al. (2016)
+/// ArXiV: <https://arxiv.org/abs/1602.05150>
+///
```

Comment:
We should also document this function is multithreaded with rayon, since a rust user might need to know this depending on the larger global threading context. I do wonder if we want to add an argument for running in parallel like we do in other rustworkx-core functions (for example `betweenness_centrality` has `parallel_threshold` to control when rayon is used or not).

Reply:
I can add the `parallel_threshold`. Looks like I can use the `CondIterator` to choose which one.